src/nvofapi/nvofapi_common.cpp

```diff
@@ -0,0 +1,218 @@
+        privData.pPrivateData = ((NV_OF_PRIV_DATA*)initParams->hPrivData)->data;
+
+        createInfo.pNext = &privData;
+
+        auto ret = m_vkCreateOpticalFlowSessionNV(m_vkDevice, &createInfo, NULL, &m_vkOfaSession);
+
+        if (ret == VK_SUCCESS) {
+            return Success();
+        }
+
+        return ErrorGeneric();
+    }
+
+    NV_OF_STATUS NvOFInstance::BindImageToSession(NvOFImage* image, VkOpticalFlowSessionBindingPointNV bindingPoint) {
+        VkImageLayout layout = VK_IMAGE_LAYOUT_GENERAL;
+
+        if (!image)
+            return ErrorGeneric();
+
+        auto ret = m_vkBindOpticalFlowSessionImageNV(m_vkDevice,
+            m_vkOfaSession,
+            bindingPoint,
+            image->ImageView(),
+            layout);
+        if (ret != VK_SUCCESS) {
+            return ErrorGeneric();
+        }
+        return Success();
+    }
+
+    NV_OF_STATUS NvOFInstance::getCaps(NV_OF_CAPS param, uint32_t* capsVal, uint32_t* size) {
+        if (param == NV_OF_CAPS_SUPPORTED_OUTPUT_GRID_SIZES) {
+            *size = 1;
+            if (capsVal) {
+                *capsVal = NV_OF_OUTPUT_VECTOR_GRID_SIZE_4;
+            }
+
+            // XXX[ljm] query VkPhysicalDevice for actual support
+            return Success();
+        }
+        return ErrorGeneric();
+    }
+
+    void NvOFInstance::RegisterBuffer(const NV_OF_REGISTER_RESOURCE_PARAMS_VK* registerParams) {
+        NvOFImage* nvOFImage = new NvOFImage(m_vkDevice, registerParams->image, registerParams->format);
+        nvOFImage->Initialize(m_vkCreateImageView, m_vkDestroyImageView);
+        *registerParams->hOFGpuBuffer = reinterpret_cast<NvOFGPUBufferHandle>(nvOFImage);
+    }
+
+    void NvOFInstance::RecordCmdBuf(const NV_OF_EXECUTE_INPUT_PARAMS_VK* inParams, NV_OF_EXECUTE_OUTPUT_PARAMS_VK* outParams, VkCommandBuffer cmdBuf) {
+        BindImageToSession(reinterpret_cast<NvOFImage*>(inParams->inputFrame), VK_OPTICAL_FLOW_SESSION_BINDING_POINT_INPUT_NV);
+        BindImageToSession(reinterpret_cast<NvOFImage*>(inParams->referenceFrame), VK_OPTICAL_FLOW_SESSION_BINDING_POINT_REFERENCE_NV);
+        BindImageToSession(reinterpret_cast<NvOFImage*>(outParams->outputBuffer), VK_OPTICAL_FLOW_SESSION_BINDING_POINT_FLOW_VECTOR_NV);
+        BindImageToSession(reinterpret_cast<NvOFImage*>(outParams->outputCostBuffer), VK_OPTICAL_FLOW_SESSION_BINDING_POINT_COST_NV);
+        BindImageToSession(reinterpret_cast<NvOFImage*>(outParams->bwdOutputBuffer), VK_OPTICAL_FLOW_SESSION_BINDING_POINT_BACKWARD_FLOW_VECTOR_NV);
+        BindImageToSession(reinterpret_cast<NvOFImage*>(outParams->bwdOutputCostBuffer), VK_OPTICAL_FLOW_SESSION_BINDING_POINT_BACKWARD_COST_NV);
+        BindImageToSession(reinterpret_cast<NvOFImage*>(outParams->globalFlowBuffer), VK_OPTICAL_FLOW_SESSION_BINDING_POINT_GLOBAL_FLOW_NV);
+        // Support INPUT_MIPS execute priv data
+        if (((NV_OF_PRIV_DATA*)inParams->hPrivData)->id == NV_OF_EXECUTE_PRIV_DATA_ID_INPUT_MIPS) {
```

Comment:
Besides the missing null check for `hPrivData`, is there no need to also check the size of what `data` points to?

Reply:
I don't think that's necessary, the size can be inferred by the type.